odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+            onedim_arr *= stride * _interp_kernel_ft(freqs, intp)
+        else:
+            onedim_arr /= stride * _interp_kernel_ft(freqs, intp)
+
+        onedim_arrs.append(onedim_arr.astype(out.dtype, copy=False))
+
+    fast_1d_tensor_mult(out, onedim_arrs, axes=axes, out=out)
+    return out
+
+
+def reciprocal_space(space, axes=None, halfcomplex=False, shift=True,
+                     **kwargs):
+    """Return the range of the Fourier transform on ``space``.
+
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        Real space whose reciprocal is calculated. It must be
+        uniformly discretized.
+    axes : `sequence` of ints, optional
+        Dimensions along which the Fourier transform is taken.
+        Default: all axes
+    halfcomplex : bool, optional
+        If ``True``, take only the negative frequency part along the last
+        axis for. For ``False``, use the full frequency space.
+        This option can only be used if ``space`` is a space of
+        real-valued functions.
+    shift : bool or `sequence` of bools, optional
+        If ``True``, the reciprocal grid is shifted by half a stride in
+        the negative direction. With a boolean sequence, this option
+        is applied separately to each axis.
+        If a sequence is provided, it must have the same length as
+        ``axes`` if supplied. Note that this must be set to ``True``
+        in the halved axis in half-complex transforms.
+        Default: ``True``
+    exponent : float, optional
+        Create a space with this exponent. By default, the conjugate
+        exponent ``q = p / (p - 1)`` of the exponent of ``space`` is
+        used, where ``q = inf`` for ``p = 1`` and vice versa.
+    dtype : optional
+        Complex data type of the reciprocal space. By default, the
+        complex counterpart of ``space.dtype`` is used.
+
+    Returns
+    -------
+    rspace : `DiscreteLp`
+        Reciprocal of the input ``space``. If ``halfcomplex=True``, the
+        upper end of the domain (where the half space ends) is chosen to
+        coincide with the grid node.
+    """
+    if not isinstance(space, DiscreteLp):
+        raise TypeError('`space` {!r} is not a `DiscreteLp` instance'
+                        ''.format(space))
+    if not space.is_uniform:
+        raise ValueError('`space` is not uniformly discretized')
+
+    if axes is None:
+        axes = list(range(space.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    if halfcomplex and space.field != RealNumbers():
+        raise ValueError('`halfcomplex` option can only be used with real '
+                         'spaces')
+
+    exponent = kwargs.pop('exponent', None)
+    if exponent is None:
+        exponent = conj_exponent(space.exponent)
+
+    dtype = kwargs.pop('dtype', None)
+    if dtype is None:
+        if is_real_dtype(space.dtype):
+            dtype = TYPE_MAP_R2C[space.dtype]
```

Comment:
This TYPE_MAP_R2C magic should simply be replaces with a utility function `complex_dtype` that works on any dtype, that way you can remove the above if statement as well.